---
title: "Preliminary analysis of MURI module 3 sequence data"
format: 
  html:
    embed-resources: true
execute:
  message: false
  warning: false
editor: visual
code-fold: true
fig-width: 15
fig-height: 8
page-layout: full
---

```{r}
#| echo: false
#| message: false
#| warning: false

rm(list = ls())
list.of.packages <- c("phyloseq", "tidyverse","pander","ggsci","ggplot2","pandoc")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

library(phyloseq)
library(tidyverse)
library(pander)
library(ggsci)
library(ggplot2)
library(pandoc)
```

## Background

This document contains preliminary analyses of the initial module 3 method selection experiment, which tests various combinations of primer and PCR conditions on the relative success of sequencing various target species from four broad taxonomic categories: cetaceans, krill, fishes (including sharks), and cephalopods.

Sequence data were generated on the NWFSC Conservation Genetics Lab MiSeq, and processed using a dada2-based QAQC pipeline. Sequence files were processed in batches, grouped by target region, using trimming and filtering parameters specific to the primer length and target region length for each batch of samples. A total of 7 output files were generated, and are store as .rdata files **HERE** and **HERE**.

This document combines the output from all batches, converts the data to the phyloseq format, and generates plots for preliminary exploration of the data.

## Pipeline assessment: read depth

The following shows the mean read depth across all samples after each major filtering step in the dada2 pipeline.

```{r}
output.files <- list.files("~/Desktop/muri_sandbox/example_data_structure/final_data/", full.names = TRUE, pattern = "*.Rdata")

output.list <- list()

for (i in 1:length(output.files)){
load(output.files[[i]],  temp_env <- new.env())
output.list[[i]] <- as.list(temp_env)
}
rm(temp_env)

track <- data.frame()

for (i in 1:length(output.files)){
temp.track <- output.list[[i]][["track"]]
track <- rbind(track, temp.track)
rm(temp.track)
}

track.sum <- track %>% 
  as.data.frame() %>% 
  summarize_all(mean)

pander(track.sum, big.mark = ",")
```

## High-level performance: Read depth by taxonomic class

The number of reads by taxonomic class for each primer and taq combination are shown below:

```{r}
#| warning: false

# make sequence table
seqtab.nochim <- as.data.frame(output.list[[1]][["seqtab.nochim"]])

for (i in 2:length(output.list)){
seqtab.nochim.temp <- as.data.frame(output.list[[i]][["seqtab.nochim"]])
seqtab.nochim <- bind_rows(seqtab.nochim, seqtab.nochim.temp)
rm(seqtab.nochim.temp)
}

seqtab.nochim <- seqtab.nochim %>% 
  mutate_all(coalesce, 0) %>% 
  as.matrix()

# get sample names
samples.out <- rownames(seqtab.nochim)

# make table of sample metadata
samdf <- samples.out %>% 
  as.data.frame(col.names = "samples") %>% 
  rename("samples" = 1) %>% 
  mutate("samples2" = samples) %>% 
  separate(samples2,into = c("primer","taq","replicate","td")) %>%
  mutate(td = case_when(grepl("S", td) ~ 0, TRUE ~ 1)) %>% 
  group_by(primer,taq,td) %>% 
  mutate(replicate = row_number()) %>% 
  ungroup() %>% 
  unite("primer.taq", primer:taq, remove = FALSE) %>% 
  column_to_rownames("samples")

# make table of taxonomic classifications

taxa <- data.frame()

for (i in 1:length(output.list)){
taxa.temp <- output.list[[i]][["taxa"]]
taxa <- rbind(taxa,taxa.temp)
rm(taxa.temp)
}

taxa <- taxa %>% rownames_to_column(var = "asv") 

tax.table <- as.data.frame(dimnames(seqtab.nochim)[[2]]) %>% 
  rename("asv" = 1) %>% 
  left_join(taxa, by = c("asv" = "asv")) %>% 
  mutate(Species = case_when(grepl("catodon", Species) ~ "Physeter macrocephalus", TRUE ~ Species)) %>% 
  separate(Species, into = c("Genus1", NA), sep = " ", remove = FALSE) %>% 
  mutate(across(where(is.character), ~na_if(.,"NA"))) %>% 
  mutate(Genus = coalesce(Genus, Genus1)) %>% 
  select(-Genus1)

tax.table.fill <- tax.table %>% 
  filter(!(is.na(Genus))) %>% 
  group_by(Genus) %>% 
  fill(c("Kingdom","Phylum","Class","Order","Family"), .direction = "downup") %>% 
  column_to_rownames("asv") %>% 
  as.matrix()

# build phyloseq
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(tax.table.fill))

# remove undetermined samples
ps <- subset_samples(ps, primer != "Undetermined")

# give asvs shorter names and store their sequences in a separate slot in the phyloseq
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))

# merge asvs by species
ps.sp <- tax_glom(ps, "Species")
sp.tax <- ps.sp@tax_table@.Data

plot_bar(ps.sp, fill="Class") +
  theme(axis.title=element_text(size = 16),
        legend.title = element_text(size = 16),
        axis.text.x=element_blank(),
        legend.text = element_text(size = 16),
        axis.ticks.x=element_blank()) +
  xlab("Primer/taq combinations") +
  facet_wrap(~primer.taq, scales = "free_x") +
  scale_fill_hue(c = 60)
```

## High-level performance: Proportional reads by taxonomic class

We can convert the number of reads to proportions and look at the same plot to see how evenly distributed reads are among taxonomic class. A few interesting things pop out here:

-   The figure is colored by taxonomic class, but individual species are still delineated by black lines within each class.

-   Note that despite large differences in reads among replicates, the proportion of reads going to each class and species seems to be very even among replicates.

-   Note a large proportion of off-target reads with the MV1 and MV3 primers, mostly from fishes and NA taxa - NA taxa could be marine mammals if we don't have sequences from that region for those taxa. We can look into that.

-   DL looks great for marine mammals! Read depth doesn't look very even across all species, suggesting significant amplification bias - or maybe lack of taxonomic resolution.

-   C18 looks great for cephalopods!

-   C16: it's interesting that the taq used seems to have had a huge effect on what taxa were sequenced here!

-   MFU has a lot of NA taxa, likely bacterial but we can look into those later. It also doesn't really pick up Chondrichthyes.

-   Nothing seems to have picked up Malacostraca

-   LRY might be a decent first pass primer - it'll pick up the fishes and sharks and give us a sense of whether or not there are marine mammals in a sample (although likely not great species-level resolution. We can take a closer look at how faithfully the read depth reflects input DNA, and what kind of taxonomic resolution we got from that primer set.

```{r species_prop_barplot}

ps.sp.prop <- transform_sample_counts(ps.sp, function(x) x/sum(x))

plot_bar(ps.sp.prop, fill="Class") +
  theme(axis.title=element_text(size = 16),
        legend.title = element_text(size = 16),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 16),
        legend.position = "right") +
  xlab("Primer/taq combinations") +
  facet_wrap(~primer.taq, scales = "free_x") +
  scale_fill_hue(c = 60)
```

## To-do:

Some first thoughts on next steps include:

-   BLAST the NA sequences to see what they are (bacteria? un-id'ed fishes pr whales? etc)

-   Look into why krill don't seem to be showing up here.

-   Double check C16 to make sure it's running as expected.

-   Group reads by target region and class, and see how species-level read depth compares with the input DNA (my understanding is that input DNA for each class is even, correct?)

-   Compare the ID'ed species to the input species, both for individual samples and across all replicates combined.

    -   True positives and False positives

## Diversity

Observed and Shannon diversity indices will give us an idea of which samples had the highest number of species and highest evenness, respectively. I don't think this means anything at the current stratification, but might be more informative once I stratify the dataset by primer and taxonomic class.

```{r}
#| label: fig-diversity
#| fig-cap: Observed species counts and Shannon diversity (evenness) indices for each sample in the primer test dataset.
plot_richness(ps.sp, x = "primer.taq", measures=c("Observed", "Shannon"), color = "primer", shape = "taq") +
  scale_color_aaas() +
  geom_point(size = 3) +
  theme(legend.text = element_text(size = 16),
        legend.title = element_text(size = 16),
        axis.text.x = element_blank(),
        axis.title = element_text(size = 16)) +
  xlab("Primer/taq combinations")
```

```{r}
#| label: data-save

save(ps, ps.sp, ps.sp.prop, file = "primer_test_phyloseq.Rdata")

```
